src/dal/views.py

```diff
@@ -92,11 +103,55 @@ def get_queryset(self):
         """Filter the queryset with GET['q']."""
         qs = super(BaseQuerySetView, self).get_queryset()
 
-        if self.q:
-            qs = qs.filter(**{'%s__icontains' % self.model_field_name: self.q})
+        qs = self.get_search_results(qs, self.q)
 
         return qs
 
+    def get_search_fields(self):
+        """Get the fields to search over."""
+        if self.search_fields:
+            return self.search_fields
+        else:
+            return [self.model_field_name]
+
+    def _construct_search(self, field_name):
+        """Apply keyword searches."""
+        if field_name.startswith("^"):
+            return "%s__istartswith" % field_name[1:]
+        elif field_name.startswith("="):
+            return "%s__iexact" % field_name[1:]
+        elif field_name.startswith("@"):
+            return "%s__search" % field_name[1:]
+        else:
+            return "%s__icontains" % field_name
+
+    def get_search_results(self, queryset, search_term):
+        """Filter the results based on the query."""
+        search_fields = self.get_search_fields()
+        if search_fields and search_term:
+            orm_lookups = [
+                self._construct_search(search_field) for search_field in search_fields
+            ]
+            if self.split_words is not None:
+                word_conditions = []
+                for word in search_term.split():
+                    or_queries = [Q(**{orm_lookup: word}) for orm_lookup in orm_lookups]
+                    word_conditions.append(reduce(operator.or_, or_queries))
+                op_ = operator.or_ if self.split_words == "or" else operator.and_
+                queryset = queryset.filter(reduce(op_, word_conditions))
+            else:
+                or_queries = [
+                    Q(**{orm_lookup: search_term}) for orm_lookup in orm_lookups
+                ]
+                queryset = queryset.filter(reduce(operator.or_, or_queries))
+
+            for search_spec in orm_lookups:
+                if lookup_needs_distinct(queryset.model._meta, search_spec):
+                    queryset = queryset.distinct()
```

Comment:
Is this calling the same distinct() for every search_spec for every search_field ?

New code by Дилян Палаузов in 2017 in Django seems to perform much better.

Reply:
It should not - once it finds one that requires distinct, the code `break`s the loop.  I have updated it to use the new code since you say it performs better.

DALv4 will be backward compatible?  I am just upgrading to v3 now, I hope I will not need to do a big update again.

Reply:
DAL V4 is compatible